File: leetcode/javascript/0016_3Sum_closest.js
// https://leetcode.com/problems/3sum-closest/
// O(n^2) time complexity
// O(log(n)) space complexity
function threeSumClosest(nums, target) {
  nums.sort((a, b) => a - b);
  let difference = Infinity;
  for (let i = 0; i < nums.length && difference !== 0; i++) {
    let [left, right] = [i + 1, nums.length - 1];
    while (left < right) {
      const sum = nums[i] + nums[left] + nums[right];
      if (Math.abs(sum - target) < Math.abs(difference)) {
        difference = sum - target;
      }
      if (sum > target) {
        right -= 1;
      } else {
        left += 1;
      }
    }
  }
  return difference + target;
}

File: leetcode/python/0016_3sum_closest.py
# https://leetcode.com/problems/3sum-closest/
# O(n^2) time complexity
# O(1) space complexity
def three_sum_closest(nums, target):
    nums.sort()
    diff = float('inf')
    for i in range(len(nums)):
        left, right = i + 1, len(nums) - 1
        while left < right:
            total = nums[i] + nums[left] + nums[right]
            if total < target:
                left += 1
            elif total > target:
                right -= 1
            else:
                return target
            diff = target - total if abs(target - total) < abs(diff) else diff
    return target - diff